File: performance-comparison/benchmarks/concurrency/python/thread_ring.py
import sys
import threading
import queue

n_threads = int(sys.argv[1]) if len(sys.argv) > 1 else 4
messages  = int(sys.argv[2]) if len(sys.argv) > 2 else 1000

chans = [queue.Queue(maxsize=1) for _ in range(n_threads)]
done  = threading.Event()

# Adjust so the ring completes at thread 0 (messages must be a multiple of n_threads)
adj = (n_threads - messages % n_threads) % n_threads
total = messages + adj

def worker(i):
    me  = chans[i]
    nxt = chans[(i + 1) % n_threads]
    while True:
        tok = me.get()
        if tok <= 0:
            if i == 0:
                done.set()
            else:
                nxt.put(tok - 1)
            return
        nxt.put(tok - 1)

threads = [threading.Thread(target=worker, args=(i,), daemon=True) for i in range(n_threads)]
for t in threads: t.start()
chans[0].put(total)
done.wait()
print("done")

File: performance-comparison/benchmarks/concurrency/clojure/thread_ring.clj
(let [n-threads  (Integer/parseInt (first *command-line-args*))
     messages    (Long/parseLong (second *command-line-args*))
     ;; adjust so token reaches 0 at thread 0
     adj         (mod (- n-threads (mod messages n-threads)) n-threads)
     total       (+ messages adj)
     chans       (vec (repeatedly n-threads #(java.util.concurrent.LinkedBlockingQueue. 1)))
     latch       (java.util.concurrent.CountDownLatch. 1)]
  (doseq [i (range n-threads)]
    (let [me   (nth chans i)
          next (nth chans (mod (inc i) n-threads))]
      (future
        (loop []
          (let [tok (.take me)]
            (cond
              (<= tok 0)
              (when (= i 0) (.countDown latch))
              :else
              (do (.put next (dec tok)) (recur))))))))
  (.put (first chans) total)
  (when-not (.await latch 30 java.util.concurrent.TimeUnit/SECONDS)
    (binding [*out* *err*]
      (println "ERROR: thread ring did not complete within 30s"))
    (System/exit 1))
  (println "done")
  (System/exit 0))
